source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -885,9 +915,26 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
 
 void Filter::onUpstreamConnection() {
   connecting_ = false;
-  // Re-enable downstream reads now that the upstream connection is established
-  // so we have a place to send downstream data to.
-  read_callbacks_->connection().readDisable(false);
+  // If receive_before_connect is not enabled or we have some early data in the buffer,
+  // means that we have previously read disabled the downstream. Now that the upstream
+  // connection is established, we should enable reads back.
+  bool should_read_enable = !receive_before_connect_ || early_data_buffer_.length() > 0;
```

Comment:
I think this code would be more clear if you flipped this conditional and returned early. Something like:
```
if receive_before_connect_ && early_data_buffer_.length() == 0 {
    return
}
if early_data_buffer_.length() > 0 {
  // flush early data
}
read_callbacks_->connection().readDisable(false);
}
```

Reply:
There is bunch of other stuff that happens in this function so I cant just return early. I can extract this part to a separate function and handle it there but not sure what to name that function.

Reply:
I can make it such that we remove the not condition as you mentioned:

```
  bool read_disabled_before = receive_before_connect_ && early_data_buffer_.length() > 0;

  // If we have received any data before upstream connection is established, send it to
  // the upstream connection.
  if (early_data_buffer_.length() > 0) {
    ENVOY_CONN_LOG(debug, "TCP:onUpstreamEvent() Flushing early data buffer to upstream",
                   read_callbacks_->connection());
    getStreamInfo().getUpstreamBytesMeter()->addWireBytesSent(early_data_buffer_.length());
    upstream_->encodeData(early_data_buffer_, early_data_end_stream_);
    ASSERT(0 == early_data_buffer_.length());
  }

  if (read_disabled_before) {
    // Re-enable downstream reads now that the upstream connection is established
    // and early data if any has been flushed to the upstream.
    read_callbacks_->connection().readDisable(false);
  }
```

What do you think ?

Reply:
Looking better, but then since you only use `read_disabled_before` in the if, just inline it?

Reply:
actually why not just nest it inside the `early_data_buffer_.length() > 0` block like:
```
/ If we have received any data before upstream connection is established, send it to
  // the upstream connection.
  if (early_data_buffer_.length() > 0) {
    ENVOY_CONN_LOG(debug, "TCP:onUpstreamEvent() Flushing early data buffer to upstream",
                   read_callbacks_->connection());
    getStreamInfo().getUpstreamBytesMeter()->addWireBytesSent(early_data_buffer_.length());
    upstream_->encodeData(early_data_buffer_, early_data_end_stream_);
    ASSERT(0 == early_data_buffer_.length());

      if (receive_before_connect_) {
        // Re-enable downstream reads now that the upstream connection is established
        // and early data if any has been flushed to the upstream.
        read_callbacks_->connection().readDisable(false);
      }
}
```

Reply:
Discussed offline, resolving for now

Reply:
Discussed offline and modified accordingly.